### PR TITLE
[mlir][ODS] Verify type constraints in Types and Attributes

### PR DESCRIPTION
**Pull request by matthias-springer**
When a type/attribute is defined in TableGen, a type constraint can be used for parameters, but the type constraint verification was missing.

Example:
```
def TestTypeVerification : Test_Type<"TestTypeVerification"> {
  let parameters = (ins AnyTypeOf<[I16, I32]>:$param);
  // ...
}
```

No verification code was generated to ensure that `$param` is I16 or I32.

When type constraints a present, a new method will generated for types and attributes: `verifyInvariantsImpl`. (The naming is similar to op verifiers.) The user-provided verifier is called `verify` (no change). There is now a new entry point to type/attribute verification: `verifyInvariants`. This function calls both `verifyInvariantsImpl` and `verify`. If neither of those two verifications are present, the `verifyInvariants` function is not generated.

When a type/attribute is not defined in TableGen, but a verifier is needed, users can implement the `verifyInvariants` function. (This function was previously called `verify`.)

Note for LLVM integration: If you have an attribute/type that is not defined in TableGen (i.e., just C++), you have to rename the verification function from `verify` to `verifyInvariants`. (Most attributes/types have no verification, in which case there is nothing to do.)

Depends on #102657.

### CODE REVIEW

File: mlir/include/mlir/IR/Types.h [hunk 1]
```diff
@@ -97,20 +97,17 @@ class Type {
...
+  [[deprecated("Use mlir::dyn_cast<U>() instead")]] U dyn_cast() const;
   template <typename U>
-  [[deprecated("Use mlir::dyn_cast_or_null<U>() instead")]]
-  U dyn_cast_or_null() const;
+  [[deprecated("Use mlir::dyn_cast_or_null<U>() instead")]] U
+  dyn_cast_or_null() const;
   template <typename U>
-  [[deprecated("Use mlir::cast<U>() instead")]]
-  U cast() const;
+  [[deprecated("Use mlir::cast<U>() instead")]] U cast() const;
```
Comment (River707) [inline mlir/include/mlir/IR/Types.h:110]: Unrelated formatting change?

File: mlir/tools/mlir-tblgen/AttrOrTypeDefGen.cpp [hunk 2]
```diff
@@ -295,24 +309,91 @@ void DefGen::emitDialectName() {
...
+                       Method::Static, builderParams);
+  verifier->body().indent();
+  if (hasImpl) {
+    // Call the verifier that checks the type constraints.
+    verifier->body() << "if (::mlir::failed(verifyInvariantsImpl(";
+    for (int i = 0, e = builderParams.size(); i < e; ++i) {
+      if (i > 0)
+        verifier->body() << ", ";
+      verifier->body() << builderParams[i].getName();
+    }
```
Comment (River707) [inline mlir/tools/mlir-tblgen/AttrOrTypeDefGen.cpp:379]: Can you use llvm::interleave(comma) here? Same below.

**Review by River707 — COMMENTED** [review]
What is the plan for attribute constraints? Are you intending to do that in a followup? Or can you just add it here? Would be good to have parity between the two.

File: mlir/tools/mlir-tblgen/AttrOrTypeDefGen.cpp (hunk 2 above)
Comment (River707) [reply mlir/tools/mlir-tblgen/AttrOrTypeDefGen.cpp:379]: Also, the below loop looks identical aside from the verify call, can you merge the two and just change the name based on the flags?

**Comment by River707** [comment]
This looks quite nice though! It's good to have more parity between ops/non-ops. Can you also add some documentation to the attr/type definition page detailing that constraints can also be used in parameters? (not sure if there is a good spot for that)

**Comment by matthias-springer** [comment]
> What is the plan for attribute constraints? Are you intending to do that in a followup? Or can you just add it here? Would be good to have parity between the two.

I'm going to take a look tmr. I think it's very similar and can be done in this PR.

File: mlir/include/mlir/IR/Constraints.td [hunk 3]
```diff
@@ -153,6 +153,9 @@ class TypeConstraint<Pred predicate, string summary = "",
     Constraint<predicate, summary> {
   // The name of the C++ Type class if known, or Type if not.
   string cppClassName = cppClassNameParam;
+  // TODO: This field is sometimes called `cppClassName` and sometimes
```
Comment (jpienaar) [inline mlir/include/mlir/IR/Constraints.td:156]: This one has cppClassName above, is this effectively duplicating it to enable reuse? (Alternative would be having ODS check for either as staging mechanism + no need to introduce here, then full rename can be done)

Comment (jpienaar) [reply mlir/include/mlir/IR/Constraints.td:156]: (having seen usage, what you have done look like less changes)

**Review by jpienaar — COMMENTED** [review]
Overall looks good! It's nice and compact :-)  River made good point and good to have together as you proposed.

**Comment by matthias-springer** [comment]
I took a look at attribute constraints. My implementation already generates the verifier, but a different part of the TableGen generator does not support it.

Example:
```
def TestAttrVerification : Test_Attr<"TestAttrVerification"> {
  let parameters = (ins AnyAttrOf<[I16Attr, I32Attr]>:$param);
  let mnemonic = "attr_verification";
  let assemblyFormat = "`<` $param `>`";
}
```

```
llvm-project/mlir/include/mlir/IR/CommonAttrConstraints.td:171:7: error: Missing `cppType` field in Attribute/Type parameter: anonymous_718
class AnyAttrOf<list<Attr> allowedAttrs, string summary = "",
```

The reason is that `AttributeConstraint` is slightly different from `TypeConstraint`: it does not have a `cppType`/`cppClassName` field. Adding that could be a larger change, so I'd rather do that in a separate commit.

I added another test to `attr-or-type-format.td` that shows that a verifier is generated for parameters that are attributes (`AttrDef`).

**Review by River707 — APPROVED** [review]

File: mlir/include/mlir/IR/Constraints.td (hunk 3 above)
Comment (matthias-springer) [reply mlir/include/mlir/IR/Constraints.td]: Fixed in #102657.
